src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java

```diff
@@ -297,15 +298,66 @@ public SegmentMetadata.ComponentMetadataMap writeData(IndexDescriptor indexDescr
                 if (vectorPostings.shouldAppendDeletedOrdinal())
                     deletedOrdinals.add(vectorPostings.getOrdinal());
             }
+
+            // map of existing ordinal to rowId (aka new ordinal if remapping is possible)
+            BiMap <Integer, Integer> ordinalMap = HashBiMap.create();
+            boolean canFastFindRows = false;
+            if (deletedOrdinals.size() == 0) {
+                canFastFindRows = true;
+                int minRow = Integer.MAX_VALUE;
+                int maxRow = Integer.MIN_VALUE;
+                for (VectorPostings<T> vectorPostings : postingsMap.values())
+                {
+                    if (vectorPostings.getRowIds().size() != 1)
+                    {
+                        canFastFindRows = false;
+                        ordinalMap.clear();
+                        break;
+                    }
+                    int rowId = vectorPostings.getRowIds().getInt(0);
+                    int ordinal = vectorPostings.getOrdinal();
+                    minRow = Math.min(minRow, rowId);
+                    maxRow = Math.max(maxRow, rowId);
+                    if(rowId != ordinal)
+                    {
+                        if (ordinalMap.containsKey(ordinal) && ordinalMap.get(ordinal) != rowId)
+                        {
+                            canFastFindRows = false;
+                            ordinalMap.clear();
+                            break;
+                        } else {
+                            ordinalMap.put(ordinal, rowId);
+                        }
+                    }
+                }
+
+                if (canFastFindRows && (minRow != 0 || maxRow != postingsMap.values().size() - 1))
+                {
+                    canFastFindRows = false;
+                    ordinalMap.clear();
+                }
+            }
+
+            Function<Integer, Integer> ordinalMapper = ordinalMap.isEmpty()
```

Comment:
don't we need to check canFastFindRows here?

Reply:
ordinalMap is empty in that case. I'll add it for clarity but really we don't have to.

Reply:
ah, you're right.  don't worry about changing it, then